core/src/main/scala/io/isomarcte/http4s/active/requests/core/ActiveRequestMiddleware.scala

```diff
@@ -0,0 +1,161 @@
+    * @param  initial the initial state value.
+    * @param  onStart a function to apply to the current state each time a new
+    *                 `Request` is received.
+    * @param  onEnd   a function to apply to the current state each time a
+    *                 `Response` is returned.
+    * @param action a function of the current state and the `Request` which
+    *                yields either a `Request` or a `Response`. If a
+    *                `Response` is yielded this has the effect of bypassing
+    *                the underlying service.
+    *
+    * @tparam F a `Sync` type.
+    * @tparam S the state.
+    *
+    * @return a pair of a `F[S]` which can be used to inspect the current
+    *         state externally and the middleware.
+    */
+  def primitive[F[_], S](
+    initial: S,
+    onStart: S => S,
+    onEnd: S => S,
+    action: (S, Request[F]) => F[Either[Request[F], Response[F]]],
+  )(
+    implicit F: Sync[F]
+  ): (F[S], HttpMiddleware[F]) = {
+    val state: AtomicReference[S] = new AtomicReference(initial)
+    val onStartOp: F[S] = F.delay(state.updateAndGet(unaryOp(onStart)))
+    val onEndOp: F[Unit] = F.delay(state.updateAndGet(unaryOp(onEnd))).void
+    def use(
+      service: Kleisli[OptionT[F, ?], Request[F], Response[F]],
+      req: Request[F]
+    )(s: S): Stream[F, Option[Response[F]]] =
+      Stream.eval(
+        action(s, req).flatMap{
+          case Left(req) =>
+            service.run(req).value
+          case Right(resp) =>
+            F.pure(Option(resp))
+        }
+      )
+
+    (F.delay(state.get),
+      (service: Kleisli[OptionT[F, ?], Request[F], Response[F]]) =>
+      Kleisli[OptionT[F, ?], Request[F], Response[F]]((req: Request[F]) =>
+        OptionT(
+          Stream.bracket(onStartOp)(
+            use = use(service, req),
+            release = Function.const(onEndOp)
+          ).compile.toList.map((l: List[Option[Response[F]]]) => l.headOption.flatten))
+      )
+    )
+  }
+
+  /** Middleware which counts the active requests and allows for an action to be
+    * taken based off the current number of active requests and the current
+    * request.
+    *
+    * @param action a function of the number of active requests and the
+    *                current `Request` which yields either a `Request` or a
+    *                `Response`. If a `Response` is yielded this has the
+    *                effect of bypassing the underlying service.
+    *
+    * @tparam F a `Sync` type.
+    * @tparam N a `Numeric` state type, e.g. `Long`.
+    *
+    * @return a pair of a `F[S]` which can be used to inspect the current
+    *         state externally and the middleware.
+    */
+  def activeRequestCountMiddleware[F[_], N](
+    action: (N, Request[F]) => F[Either[Request[F], Response[F]]]
+  )(
+    implicit F: Sync[F],
+             N: Numeric[N]
+  ): (F[N], HttpMiddleware[F]) =
+    this.primitive(N.zero, (n: N) => N.plus(n, N.one), (n: N) => N.minus(n, N.one), action)
```

Comment:
No, but `Long` will get boxed too. If I used `AtomicLong` then we may avoid boxing, but given the cardinality of these will likely never break 100, I'm not worried about it. That is, I'd rather have a generic state, the attempt to avoid boxing. If we use `AtomicLong` we obviously loose generic state (unless I just write two function :P)